dbms/src/Compression/CompressionCodecDoubleDelta.cpp

```diff
@@ -0,0 +1,320 @@
+#include <IO/BitHelpers.h>
+#include <IO/WriteHelpers.h>
+
+#include <string.h>
+#include <algorithm>
+#include <cstdlib>
+#include <type_traits>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int CANNOT_COMPRESS;
+extern const int CANNOT_DECOMPRESS;
+}
+
+namespace
+{
+UInt32 getDeltaTypeByteSize(UInt8 data_bytes_size)
+{
+    // both delta and double delta can be twice the size of data item, but not less than 32 bits and not more that 64.
+    return std::min(64/8, std::max(32/8, data_bytes_size * 2));
+}
+
+UInt32 getCompressedHeaderSize(UInt8 data_bytes_size)
+{
+    const UInt8 items_count_size = 4;
+
+    return items_count_size + data_bytes_size + getDeltaTypeByteSize(data_bytes_size);
+}
+
+UInt32 getCompressedDataSize(UInt8 data_bytes_size, UInt32 uncompressed_size)
+{
+    const UInt32 items_count = uncompressed_size / data_bytes_size;
+
+    // 11111 + max 64 bits of double delta.
+    const UInt32 max_item_size_bits = 5 + getDeltaTypeByteSize(data_bytes_size) * 8;
+
+    // + 8 is to round up to next byte.
+    return (items_count * max_item_size_bits + 8) / 8;
+}
+
+template <typename T, typename DeltaType>
+UInt32 compressDataForType(const char * source, UInt32 source_size, char * dest)
+{
+    if (source_size % sizeof(T) != 0)
+        throw Exception("Cannot compress, data size " + toString(source_size) + " is not aligned to " + toString(sizeof(T)), ErrorCodes::CANNOT_COMPRESS);
+    const char * source_end = source + source_size;
+
+    const UInt32 items_count = source_size / sizeof(T);
+    unalignedStore(dest, items_count);
+    dest += sizeof(items_count);
+
+    T prev_value{};
+    DeltaType prev_delta{};
+
+    if (source < source_end)
+    {
+        prev_value = unalignedLoad<T>(source);
+        unalignedStore(dest, prev_value);
+
+        source += sizeof(prev_value);
+        dest += sizeof(prev_value);
+    }
+
+    if (source < source_end)
+    {
+        const T curr_value = unalignedLoad<T>(source);
+        prev_delta = static_cast<DeltaType>(curr_value - prev_value);
+        unalignedStore(dest, prev_delta);
+
+        source += sizeof(curr_value);
+        dest += sizeof(prev_delta);
+        prev_value = curr_value;
+    }
+
+    WriteBuffer buffer(dest, getCompressedDataSize(sizeof(T), source_size - sizeof(T)*2));
+    BitWriter writer(buffer);
+
+    while (source < source_end)
+    {
+        const T curr_value = unalignedLoad<T>(source);
+        source += sizeof(curr_value);
+
+        const auto delta = curr_value - prev_value;
+        const DeltaType double_delta = static_cast<DeltaType>(delta - static_cast<T>(prev_delta));
+
+        prev_delta = delta;
+        prev_value = curr_value;
+
+        if (double_delta == 0)
+        {
+            writer.writeBits(1, 0);
+        }
+        else
+        {
+            const auto sign = std::signbit(double_delta);
+            const auto abs_value = static_cast<typename std::make_unsigned<DeltaType>::type>(std::abs(double_delta));
```

Comment:
```
using UnsignedDeltaType = typename std::make_unsigned<DeltaType>::type;
```
in front of function

Reply:
Also what about INT_MIN? abs(INT_MIN) == INT_MIN.

Reply:
Please take a look at the unit tests I've added. `dbms/src/Compression/tests/gtest_compressionCodec.cpp`